File: dart/lib/src/instructions/realloc_instruction.dart
import 'dart:typed_data';

import 'instruction_types.dart';

/// Realloc instruction for changing domain space allocation
///
/// This mirrors js-kit/src/instructions/reallocInstruction.ts
class ReallocInstruction extends SnsInstruction {
  ReallocInstruction({
    required this.space,
  });

  /// Instruction tag
  final int tag = 4;

  /// New space allocation
  final int space;

  String? _programAddress;
  String? _systemProgramId;
  String? _payerKey;
  String? _nameAccountKey;
  String? _nameOwnerKey;

  /// Set the parameters for the instruction
  void setParams({
    required String programAddress,
    required String systemProgramId,
    required String payerKey,
    required String nameAccountKey,
    required String nameOwnerKey,
  }) {
    _programAddress = programAddress;
    _systemProgramId = systemProgramId;
    _payerKey = payerKey;
    _nameAccountKey = nameAccountKey;
    _nameOwnerKey = nameOwnerKey;
  }

  @override
  Uint8List serialize() {
    final spaceBytes = Uint8List(4);
    spaceBytes.buffer.asByteData().setUint32(0, space, Endian.little);

    return Uint8List.fromList([
      tag,
      ...spaceBytes,
    ]);
  }

  @override
  TransactionInstruction build() {
    if (_programAddress == null ||
        _systemProgramId == null ||
        _payerKey == null ||
        _nameAccountKey == null ||
        _nameOwnerKey == null) {
      throw Exception(
          'Parameters not set for ReallocInstruction. Call setParams() first.');
    }

    final accounts = [
      AccountMeta(
        address: _systemProgramId!,
        role: AccountRole.readonly,
      ),
      AccountMeta(
        address: _payerKey!,
        role: AccountRole.writableSigner,
      ),
      AccountMeta(
        address: _nameAccountKey!,
        role: AccountRole.writable,
      ),
      AccountMeta(
        address: _nameOwnerKey!,
        role: AccountRole.readonlySigner,
      ),
    ];

    return TransactionInstruction(
      programAddress: _programAddress!,
      accounts: accounts,
      data: serialize(),
    );
  }

  /// Build the transaction instruction
  TransactionInstruction getInstruction({
    required String programAddress,
    required String systemProgramId,
    required String payerKey,
    required String nameAccountKey,
    required String nameOwnerKey,
  }) {
    setParams(
      programAddress: programAddress,
      systemProgramId: systemProgramId,
      payerKey: payerKey,
      nameAccountKey: nameAccountKey,
      nameOwnerKey: nameOwnerKey,
    );
    return build();
  }
}

File: js-kit/src/instructions/reallocInstruction.ts
import { AccountMeta, AccountRole, Address, Instruction } from "@solana/kit";
import { serialize } from "borsh";

/**
 * Input for reallocating an SNS name-registry account.
 *
 * @example
 * ```ts
 * const params: ReallocInstructionParams = { space: 1_000 };
 * ```
 */
export interface ReallocInstructionParams {
  /** New account data size in bytes. */
  space: number;
}

/** Builder for reallocating an SNS name-registry account. */
export class ReallocInstruction {
  /** Instruction discriminator. */
  tag: number;
  /** New account data size in bytes. */
  space: number;

  static schema = {
    struct: {
      tag: "u8",
      space: "u32",
    },
  };

  constructor(obj: ReallocInstructionParams) {
    this.tag = 4;
    this.space = obj.space;
  }

  serialize(): Uint8Array {
    return serialize(ReallocInstruction.schema, this);
  }

  getInstruction(
    programAddress: Address,
    systemProgramId: Address,
    payerKey: Address,
    nameAccountKey: Address,
    nameOwnerKey: Address
  ): Instruction {
    const data = this.serialize();

    const accounts: AccountMeta[] = [
      {
        address: systemProgramId,
        role: AccountRole.READONLY,
      },
      {
        address: payerKey,
        role: AccountRole.WRITABLE_SIGNER,
      },
      {
        address: nameAccountKey,
        role: AccountRole.WRITABLE,
      },
      {
        address: nameOwnerKey,
        role: AccountRole.READONLY_SIGNER,
      },
    ];

    return {
      programAddress,
      accounts,
      data,
    };
  }
}
